parlai/utils/typing.py

```diff
@@ -22,3 +24,8 @@ class _Shared(Dict[str, Any]):
 
 
 TShared = TypeVar('TShared', bound=_Shared)
+
+# ParlAI type to represent an object that is theoretically expressible as a scalar
```

Comment:
nit: could you perhaps follow the convention above of defining a class `_Scalar`? The reason for this is to have cleaner docstrings for these types (i.e. instead of `#` comments). Then you could have `TScalar = TypeVar('Tscalar', bound=_Scalar)`

Reply:
Yeah it's interesting. The typevar approach given here means that it can be used to indicate it's one of these types, but it's FIXED, and that cannot change. E.g. two references to TScalar in the same context have to either both be floats or both be ints or both be tensors.

I think for metrics we especially want to be able to mix/match so the pseudoclass which is a Union might be best. Also this excellent comment can then become a docstring.

Reply:
Lemme know what you think about my current solution, which is to just create an alias of `Union[]` rather than a subclass, because you can't subclass `Union`. See https://docs.python.org/3/library/typing.html#user-defined-generic-types